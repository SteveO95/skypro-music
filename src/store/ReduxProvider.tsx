'use client';
// Провайдер важный *элемент* который оборачивает приложение и даёт ему доступ к Хранилищу | работать ничего не будет
// Провайдер - это клиентский Компонент (к.к)
import { useRef } from 'react';
import { Provider } from 'react-redux';
import { AppStore, makeStore } from './store';
// Для того что бы вкладывать в К.К серверные компоненты,нужно использовать -children

type ReduxProviderType = {
	children: React.ReactNode;
};

export default function ReduxProvider({ children }: ReduxProviderType) {
	// storeRef- нужен для сохранения переменной между Перерендринга (перерисовок)(что бы наше значение не сбрасывалось)
	const storeRef = useRef<AppStore>();
	if (!storeRef.current) {
		storeRef.current = makeStore();
	}
	// Provider -встроенный компонент(redux.react)в него передаем *Store*-в нём храним (storeRef.current) и children это будет наша страница
	return <Provider store={storeRef.current}>{children}</Provider>;
}
