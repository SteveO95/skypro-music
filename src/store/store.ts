import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { playlistReducer } from "./features/playlistSlice";
import { userReducer } from "./features/userSlice";

export const makeStore = () => {
  return configureStore({
    reducer: combineReducers({
      playlist: playlistReducer,
      user: userReducer,
    }),
  });
};

export type AppStore = ReturnType<typeof makeStore>;
export type RootState = ReturnType<AppStore["getState"]>;
export type AppDispatch = AppStore["dispatch"];
