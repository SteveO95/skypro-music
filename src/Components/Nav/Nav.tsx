'use client'

import React from 'react'
import styles from './Nav.module.css'
import Image from 'next/image'
import Link from 'next/link'
import { routes } from '@/lib/routes'
import { useAppDispatch, useAppSelector } from '@/store/store'
import { useOutsideClick } from '@/hooks/useOutsideClick'
import { resetFilterOptions } from '@/store/features/playlistSlice'

export default function Nav() {
	const dispatch = useAppDispatch()
	const { visible, setVisible, ref } = useOutsideClick(false)
	const { user } = useAppSelector(state => state.user)
	const handleOpenMenu = () => setVisible(prev => !prev)

	return (
		<nav className={styles.main__nav}>
			<div className={styles.nav__logo}>
				<Link href={routes.HOME}>
					<Image
						className={styles.logo__image}
						src='/img/logo.png'
						alt='logo'
						width={113}
						height={17}
					/>
				</Link>
			</div>

			<div
				data-testid='burgerBtn'
				className={styles.nav__burger}
				ref={ref}
				onClick={handleOpenMenu}
			>
				<span className={styles.burger__line} />
				<span className={styles.burger__line} />
				<span className={styles.burger__line} />
			</div>

			{visible && (
				<div className={styles.nav__menu}>
					<ul className={styles.menu__list}>
						<li className={styles.menu__item}>
							<Link className={styles.menu__link} href={routes.HOME}>
								Главное
							</Link>
						</li>
						{user && (
							<li
								className={styles.menu__item}
								onClick={() => dispatch(resetFilterOptions())}
							>
								<Link className={styles.menu__link} href={routes.FAVORITE}>
									Мой плейлист
								</Link>
							</li>
						)}
						<li className={styles.menu__item}>
							<Link className={styles.menu__link} href={routes.LOGIN}>
								Войти
							</Link>
						</li>
					</ul>
				</div>
			)}
		</nav>
	)
}
