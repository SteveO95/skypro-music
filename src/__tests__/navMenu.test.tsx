import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'
import { fireEvent, render, screen } from '@testing-library/react'
import Nav from '@/components/Nav/Nav'
import { useAppDispatch, useAppSelector } from '@/store/store'

jest.mock('../store/store.ts')

const mockUseAppDispatch = useAppDispatch as jest.Mock
const mockUseAppSelector = useAppSelector as jest.Mock
const mockDispatch = jest.fn()

describe('Меню навигации', () => {
	beforeEach(() => {
		mockUseAppDispatch.mockReturnValue(mockDispatch)
		mockUseAppSelector.mockReturnValue(mockDispatch)
	})
	afterEach(() => jest.clearAllMocks())

	it('рендер компонента', () => {
		const nav = renderer.create(<Nav />).toJSON()
		expect(nav).toMatchSnapshot()
	})
	it('отображение списка меню при клике', () => {
		render(<Nav />)
		const burger = screen.getByTestId('burgerBtn')

		fireEvent.click(burger)
		const mainLink = screen.getByText('Главное')
		const authLink = screen.getByText('Войти')

		expect(mainLink).toBeInTheDocument()
		expect(authLink).toBeInTheDocument()
	})
})
