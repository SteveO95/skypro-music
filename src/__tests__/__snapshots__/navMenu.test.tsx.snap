// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Меню навигации рендер компонента 1`] = `
<nav
  className="main__nav"
>
  <div
    className="nav__logo"
  >
    <a
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      <img
        alt="logo"
        className="logo__image"
        data-nimg="1"
        decoding="async"
        height={17}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fimg%2Flogo.png&w=256&q=75"
        srcSet="/_next/image?url=%2Fimg%2Flogo.png&w=128&q=75 1x, /_next/image?url=%2Fimg%2Flogo.png&w=256&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={113}
      />
    </a>
  </div>
  <div
    className="nav__burger"
    data-testid="burgerBtn"
    onClick={[Function]}
  >
    <span
      className="burger__line"
    />
    <span
      className="burger__line"
    />
    <span
      className="burger__line"
    />
  </div>
</nav>
`;
